File: src/Enterprise/VB.NET/FlyweightExample.vb
Imports System
Imports System.Collections.Generic

Public NotInheritable Class TextStyle
    Public ReadOnly Property Font As String
    Public ReadOnly Property Size As Integer
    Public ReadOnly Property Color As String

    Public Sub New(font As String, size As Integer, color As String)
        Me.Font = font
        Me.Size = size
        Me.Color = color
    End Sub
End Class

Public NotInheritable Class StyleFactory
    Private ReadOnly styles As New Dictionary(Of String, TextStyle)()

    Public Function [Get](font As String, size As Integer, color As String) As TextStyle
        Dim key = $"{font}|{size}|{color}"
        Dim style As TextStyle = Nothing
        If Not styles.TryGetValue(key, style) Then
            style = New TextStyle(font, size, color)
            styles(key) = style
        End If
        Return style
    End Function

    Public ReadOnly Property Count As Integer
        Get
            Return styles.Count
        End Get
    End Property
End Class

Module Program
    Sub Main()
        Dim factory = New StyleFactory()
        Dim red1 = factory.Get("Inter", 12, "red")
        Dim red2 = factory.Get("Inter", 12, "red")
        factory.Get("Inter", 12, "blue")
        Console.WriteLine($"styles={factory.Count};shared={Object.ReferenceEquals(red1, red2).ToString().ToLowerInvariant()};text=ABC")
    End Sub
End Module

File: src/Enterprise/Java/FlyweightExample.java
import java.util.*;

public class FlyweightExample {
  record TextStyle(String font, int size, String color) {}
  static final class StyleFactory {
    private final Map<String, TextStyle> styles = new HashMap<>();
    TextStyle get(String font, int size, String color) {
      return styles.computeIfAbsent(font + "|" + size + "|" + color, k -> new TextStyle(font, size, color));
    }
    int count() { return styles.size(); }
  }
  record Glyph(char character, int position, TextStyle style) {}
  public static void main(String[] args) {
    var f = new StyleFactory(); var r1=f.get("Inter",12,"red"); var r2=f.get("Inter",12,"red"); var b=f.get("Inter",12,"blue");
    var g=List.of(new Glyph('A',1,r1),new Glyph('B',2,r2),new Glyph('C',3,b));
    System.out.println("styles="+f.count()+";shared="+(g.get(0).style()==g.get(1).style())+";text=ABC");
  }
}
